flink-staging/flink-hcatalog/src/main/scala/org/apache/flink/hcatalog/scala/HCatOutputFormat.scala

```diff
@@ -0,0 +1,449 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.hcatalog.scala
+
+import java.io.IOException
+import java.util
+
+import org.apache.flink.api.common.typeinfo.TypeInformation
+import org.apache.flink.api.java.typeutils.TupleTypeInfo
+import org.apache.flink.api.scala.createTypeInformation
+import org.apache.flink.api.scala.typeutils.CaseClassTypeInfo
+import org.apache.flink.hcatalog.HCatOutputFormatBase
+import org.apache.hadoop.conf.Configuration
+import org.apache.hive.hcatalog.data.{DefaultHCatRecord, HCatRecord}
+
+import scala.collection.JavaConversions.mapAsJavaMap
+import scala.collection.JavaConverters._
+
+/*
+ * A OutputFormat to write to HCatalog tables for scala api
+ * The outputFormat supports write to table and partitions through partitionValue map.
+ *
+ * scala tuple types and {@link HCatRecord} are accepted
+ * as data source. If user want to write other type of data to HCatalog, a preceding Map function
+ * is required to convert the user data type to {@link HCatRecord}
+ *
+ * The constructor provides type checking and requires import org.apache.flink.api.scala._ in the
+ * import statement of the user's program when compiling to allow scala macro to extract type info
+ */
+/**
+ *
+ * @param database hive database name, if null, default database is used
+ * @param table hive table name
+ * @param partitionValues Map of partition values, if null, whole table is used
+ * @param config the configuration
+ * @tparam T
+ */
+class HCatOutputFormat[T: TypeInformation](
+                                            database: String,
+                                            table: String,
+                                            partitionValues: Map[String, String],
+                                            config: Configuration
+                                            ) extends
+HCatOutputFormatBase[T](database, table, mapAsJavaMap(partitionValues), config) {
+
+  //get the TypeInformation
+  val ti: TypeInformation[T] = createTypeInformation[T]
+
+
+  //check the types of columns if the input is tuple(case class) type
+  if (ti.isInstanceOf[CaseClassTypeInfo[_]]) {
+    //check the number of columns
+    if (ti.asInstanceOf[CaseClassTypeInfo[_]].getArity != this.reqType.getArity) {
+      throw new IOException("tuple has different arity from the table's column numbers")
+    }
+
+    for (i <- 0 until this.reqType.getArity) {
+      val it = ti.asInstanceOf[CaseClassTypeInfo[_]].getTypeAt(i)
+      val ot = reqType.asInstanceOf[TupleTypeInfo[_]].getTypeAt(i)
+
+      if (!isCompatibleType(it, ot)) {
+        throw new IOException("field has different type from required")
+      }
+    }
+  }
```

Comment:
`isInstanceOf` is not Scala style. For this case, pattern matching would be better. I added an exception when the type of record is not derived from Tuple class.

```scala
ti match {
    case value: CaseClassTypeInfo[_] =>
      //check the number of columns
      if (value.getArity != this.reqType.getArity) {
        throw new IOException("tuple has different arity from the table's column numbers")
      }

      for (i <- 0 until this.reqType.getArity) {
        val it = value.getTypeAt(i)
        val ot = reqType.asInstanceOf[TupleTypeInfo[_]].getTypeAt(i)

        if (!isCompatibleType(it, ot)) {
          throw new IOException("field has different type from required")
        }
      }
    case _ => throw new IOException("HCatOutputFormat only supports tuple-derived type.")
  }
```